### DISCUSSION TITLE
Fresh Deno Authentication using Pocketbase

### DISCUSSION THREAD

**Discussion by RA9** [pocketbase/pocketbase#1178] [Q&A]
Hi, I was working on a project using [Deno fresh framework](https://fresh.deno.dev) and I wanted to add authentication.

As usual, I keep getting it wrong. This is the [tutorial](https://deno.com/blog/setup-auth-with-fresh) I am following so far to get this done.

My problem is, It can authenticate me but the loadFromCookie doesn't work after all. I am guessing of doing something wrong but don't know why?

**Comment by RA9** [comment]
Below is my `login.ts` file

```js
import {Handlers} from "$fresh/server.ts";
import {setCookie} from "$std/http/cookie.ts";
import {encode} from "$std/encoding/base64.ts"

export const handler: Handlers<any, { pb: any }> = {
    async POST(req, ctx) {
        const url = new URL(req.url);
        const form = await req.formData();
        const username = form.get("username");
        const password = form.get("password");

        if (username !== "" && password !== "") {
            const {token, record} = await ctx.state.pb.collection('users').authWithPassword(username, password);

            console.log({token, record})
            console.log("Is valid",ctx.state.pb.authStore.isValid);
            // return new Response('Success...');
            const headers = new Headers();
            setCookie(headers, {
                name: "pb_auth",
                value: encode(JSON.stringify({
                    token, record
                })), // this should be a unique value for each session
                maxAge: (60 * 60),
                sameSite: "Strict", // this is important to prevent CSRF attacks
                domain: url.hostname,
                path: "/",
                httpOnly: true,
                // secure: true,
            });

            // ctx.state.pb.authStore.save(token, record)

            headers.set("location", "/");
            return new Response(null, {
                status: 303, // "See Other"
                headers,
            });
        } else {
            return new Response(null, {
                status: 403,
            });
        }
    },
};
```

**Comment by RA9** [comment]
And this is my middleware file

```js
import { MiddlewareHandlerContext } from "$fresh/server.ts";
import { getCookies, setCookie } from "$std/http/mod.ts";
import PocketBase, {BaseAuthStore} from 'pocketbase';
import {config} from "$std/dotenv/mod.ts";

const configData = await config();

interface State {
  pb: {
    authStore: BaseAuthStore;
    collection: any;
  };
}


export async function handler(
  req: Request,
  ctx: MiddlewareHandlerContext<State>,
) {
  ctx.state.pb = new PocketBase(configData['PB_API_URL'] || Deno.env.get('PB_API_URL'));
  const isCookies = getCookies(req.headers);
  const cookies = req.headers.get("cookie");
  console.log("pb_auth", cookies)
  ctx.state.pb.authStore.loadFromCookie(cookies || "")
  // console.log("cookies", ctx.state.pb.authStore)
  // const isCookies = getCookies(req.headers);
  const resp = await ctx.next();

  try {
    // get an up-to-date auth store state by veryfing and refreshing the loaded auth model (if any)
    console.log("Small Sense", ctx.state.pb.authStore.isValid)
    ctx.state.pb.authStore.isValid && await ctx.state.pb.collection('users').authRefresh();
  } catch (_) {
    // clear the auth store on failed refresh
    ctx.state.pb.authStore.clear();
  }

  console.log("cookies", JSON.stringify(ctx.state.pb.authStore.exportToCookie()))


  // resp.headers.set("set-cookie", ctx.state.pb.authStore.exportToCookie())


  // setCookie(resp.headers, {
  //   name: "set-cookie",
  //   value: ctx.state.pb.authStore.exportToCookie()})

  return resp;
}
```

**Reply by RA9** [reply]
I just want to say, I am getting the `pb_auth` cookie but I don't know why it doesn't seem to work on the `authStore.loadFromCookie` 
NB: I am using the native form submission action to send the data to the `login.ts` endpoint

**Reply by ganigeorgiev** [reply]
> I am getting the pb_auth cookie but I don't know why it doesn't seem to work on the authStore.loadFromCookie

I guess this is because you are base64 encoding the cookie in login.ts. You can either try decoding it before passing to `loadFromCookie` OR remove the encoding altogether (unless it is required by setCookie, you don't really need to base64 encode the stringified json).

**Reply by RA9** [reply]
It worked after replacing the encoding with `btoa()` and `atob()`. Now my problem lies with the `ctx.state.pb.authStore.exportToCookie()` it is giving the like what `req.headers.get('cookie')` returns. How can I pass this value using the `setCookie` method?

**Reply by ganigeorgiev** [reply]
@RA9 But why do you need the base64 encoding in the first place?

> How can I pass this value using the setCookie method?

If you want to use `setCookie` method then you have to export manually the cookie (similar to how you are doing it in login.ts). You can get the `authStore` data by accessing `authStore.token` and `authStore.model`.

**Reply by RA9** [reply]
> @RA9 But why do you need the base64 encoding in the first place?
> 
I need it because `setCookie` doesn't allow a regular string as a value. They need a base64 string

**Reply by RA9** [reply]
> If you want to use `setCookie` method then you have to export manually the cookie (similar to how you are doing it in login.ts). You can get the `authStore` data by accessing `authStore.token` and `authStore.model`.

Thanks, that will work but the `Max-Age` won't be the same. Can you just tell me the default that `exportToCookie` is using?

**Reply by ganigeorgiev** [reply]
> I need it because setCookie doesn't allow a regular string as a value. They need a base64 string

From the deno documentation of useCookie and the second `cookie` argument it is not mentioned anywhere that the string needs to be base64 encoded - https://deno.land/std@0.167.0/http/mod.ts?s=Cookie. It should work with plain strings just fine but I haven't tested it so I cannot confirm.

> Can you just tell me the default that exportToCookie is using?

It is using the exp claim from the token payload and sets the expires claim (not max-age). You can use the helper `getTokenPayload()` to parse it, something like:

```js
import PocketBase, { getTokenPayload } from "pocketbase";

// extract the token expiration date
let expires;
const payload = getTokenPayload(token);
if (payload?.exp) {
    expires = new Date(payload.exp * 1000);
} else {
    expires = new Date('1970-01-01');
}
```

**Reply by RA9** [reply]
Thanks, this will work.
